src/Analyser/MutatingScope.php

```diff
@@ -678,6 +680,13 @@ private function getNodeKey(Expr $node): string
 
 	private function resolveType(string $exprString, Expr $node): Type
 	{
+		foreach ($this->expressionTypeResolverExtensionRegistry->getExtensions() as $extension) {
+			$type = $extension->getType($node);
```

Comment:
I think I need to pass `$this` here, because for [my usecase](https://github.com/phpstan/phpstan-doctrine/issues/499) I need to be able to detect the return type of the method call.

The [original solution](https://github.com/phpstan/phpstan-doctrine/pull/500/files#diff-8ee6d9610948c4e8345cb982486a7bc2a6546b7d2451966fd9df79e07042ae9eR37) used MethodReflection, but that is not possible here as the caller is unknown.

WDYT @ondrejmirtes ?

Reply:
Yeah sure, let's add Scope.

Reply:
Hm, that easily leads to infinite recursion :thinking:

Reply:
Well you shouldn't run into infinite recursion. For the same reason this one https://apiref.phpstan.org/1.11.x/PHPStan.PhpDoc.TypeNodeResolverExtension.html also does not run into infinite recursion, with basically the same design.

You should `return null;` for the nodes you are not interested in.

For method calls you should do `$scope->getMethodReflection($scope->getType($methodCall->var), $methodCall->name)`. And continue from there.

I don't think you will run into infinite recursion that way.

Reply:
Well, ok, that works. But I think most people would just use sth like `$scope->getType($methodCall)`.

So I changed the test to use this approach.